File: Week_08/java/src/BucketSort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class BucketSort {
    public static void main(String[] args) {
        int[] a1 = {1, 5, 3, 2, 6, 4};
        bucketSort(a1);
        System.out.println(Arrays.toString(a1));
    }

    public static void bucketSort(int[] array) {
        int bucketSize = 3;
        int max = array[0];
        int min = array[0];
        for (int i = 1; i < array.length; i++) {
            if (array[i] > max) {
                max = array[i];
            }
            if (array[i] < min) {
                min = array[i];
            }
        }

        //创建桶
        int bucketCount = (max - min) / bucketSize + 1;
        List[] buckets = new ArrayList[bucketCount];
        for (int i = 0; i < bucketCount; i++) {
            buckets[i] = new ArrayList();
        }

        //将元素添加到所属的桶中
        for (int i = 0; i < array.length; i++) {
            int value = array[i];
            buckets[(value - min) / bucketSize].add(value);
        }

        int k = 0;
        for (int i = 0; i < buckets.length; i++) {
            //对桶进行排序
            Collections.sort(buckets[i]);
            //排序后在取出放到原来的数组中
            for (int j = 0; j < buckets[i].size(); j++) {
                array[k++] = (int) buckets[i].get(j);
            }
        }
    }
}

File: Week_08/cpp/src/BucketSort.cpp
#include <vector>
#include <iostream>
using namespace std;
void bucketSort(vector<int> &nums)
{
    // 注意，修改nums的元素, *pmax是动态变化的
    const auto [pmin, pmax] = minmax_element(nums.begin(), nums.end());
    const int min = *pmin, max = *pmax;
    const int bucketSize = 3, bucketCount = (max - min) / bucketSize + 1;
    // 创建桶
    vector<vector<int>> buckets(bucketCount, vector<int>());
    // 将元素映射到桶中
    for (int num : nums)
        buckets[(num - min) / bucketSize].push_back(num);
    // 对每一个桶中的元素进行排序，然后取出桶中的元素
    for (int i = 0, j = 0; i < bucketCount; i++)
    {
        auto bucket = buckets[i];
        sort(bucket.begin(), bucket.end());
        for (int v : bucket)
        {
            nums[j++] = v;
        }
    }
}
int main(int argc, char const *argv[])
{
    vector<int> nums = {1, 5, 3, 2, 2, 1, 3, 6, 4};
    bucketSort(nums);
    cout << "[";
    for (int i = 0; i < nums.size(); i++)
    {
        cout << nums[i] << ",";
    }
    cout << "]" << endl;
    return 0;
}
